src/dftbp/timedep/timeprop.F90

```diff
@@ -1638,17 +1763,21 @@ subroutine kickDM(this, rho, Ssqr, Sinv, iSquare, coord)
       call gemm(rho(:,:,iKS), Sinv(:,:,iKS), T2, cmplx(0.5, 0, dp), cmplx(1, 0, dp), 'N', 'C')
     end do
 
+#:endif
     write(stdout,"(A)")'Density kicked along ' // localDir(this%currPolDir) //'!'
 
   end subroutine kickDM
 
 
   !> Creates array for an external TD field
-  subroutine getTDFunction(this, startTime)
+  subroutine getTDFunction(this, env, startTime)
```

Comment:
This routine makes too many things IMO (it did already in the serial case, but in the MPI-one it get's really convoluted):
* Calculates or reads the TD-function
* Writes the TD-function
Note, that the outcome in the parallel case is different on different ranks:
* On lead this%tdfunction is filled up with the read value, but on the other ranks, the output is undeterministic, as far as I can see.

Reply:
I see the third point you mention happening only in the the case that `tEnvFromFile` is true. In that lead knows tabout tdfun, and the others don't. But for all other conditions, tdfun is calculated on all ranks. But you are right, this function should be split at least in two cases, I'll do some refactoring now.